### PR TITLE
Republish the docker dind image

### PR DESCRIPTION
**Pull request by Piszmog**
A customer has asked for mirroring to save on long process of reviewing Docker's published image. 

## Test plan

Ran locally to check if image builds.

### CODE REVIEW

**Review by eseliger — APPROVED** [review]

**Comment by eseliger** [comment]
Might want to run a `main-dry-run` build to see what trivy thinks of this image.

**Review by burmudar — APPROVED** [review]
Thanks!

**Comment by willdollman** [comment]
Spotted Erik's comment after running a scan myself 😄. This image does have a number of vulns in it, see below. There's nothing wrong with us hosting it if it makes something easier for one customer, but we're planning on shipping it as part of sourcegraph we might need to be more conservative as any vulns would need to be added to the [accepted CVEs](https://handbook.sourcegraph.com/departments/security/tooling/trivy/4-5-0/) page.

Not sure if the customer specifically wants this version, but the latest is 23.0.1-dind. Would this need to be kept updated?

```
$ trivy image docker:20.10.22-dind@sha256:03f2d563100b9776283de1e18f10a1f0b66d2fdc7918831bf8db1cda767d6b37

Total: 28 (UNKNOWN: 0, LOW: 0, MEDIUM: 3, HIGH: 22, CRITICAL: 3)
[...]
```

**Comment by eseliger** [comment]
that tag looks _better_

```
Total: 3 (UNKNOWN: 0, LOW: 1, MEDIUM: 1, HIGH: 1, CRITICAL: 0)

┌──────────────────────────────────┬────────────────┬──────────┬───────────────────┬───────────────┬───────────────────────────────────────────────────────────┐
│             Library              │ Vulnerability  │ Severity │ Installed Version │ Fixed Version │                           Title                           │
├──────────────────────────────────┼────────────────┼──────────┼───────────────────┼───────────────┼───────────────────────────────────────────────────────────┤
│ github.com/containerd/containerd │ CVE-2023-25173 │ HIGH     │ v1.6.16           │ 1.6.18        │ containerd: Supplementary groups are not set up properly  │
│                                  │                │          │                   │               │ https://avd.aquasec.com/nvd/cve-2023-25173                │
├──────────────────────────────────┼────────────────┼──────────┼───────────────────┼───────────────┼───────────────────────────────────────────────────────────┤
│ github.com/containerd/containerd │ CVE-2023-25153 │ MEDIUM   │ v1.6.16           │ 1.6.18        │ containerd is an open source container runtime. Before    │
│                                  │                │          │                   │               │ versions 1.6.18 ...                                       │
│                                  │                │          │                   │               │ https://avd.aquasec.com/nvd/cve-2023-25153                │
├──────────────────────────────────┼────────────────┼──────────┼───────────────────┼───────────────┼───────────────────────────────────────────────────────────┤
│ golang.org/x/net                 │ CVE-2022-41723 │ LOW      │ v0.4.0            │ 0.7.0         │ A maliciously crafted HTTP/2 stream could cause excessive │
│                                  │                │          │                   │               │ CPU consumpt ...                                          │
│                                  │                │          │                   │               │ https://avd.aquasec.com/nvd/cve-2022-41723                │
└──────────────────────────────────┴────────────────┴──────────┴───────────────────┴───────────────┴───────────────────────────────────────────────────────────┘
```

but still one HIGH CVE, not sure if that's a problem for the customer :/

**Comment by alexfogg** [comment]
Thank you for this!

For CVE context- high & critical both require an approvals process at certain customers- so if there's any way to avoid them it would be super helpful to get it into their hands faster

**Comment by eseliger** [comment]
Took a look at this, it's only the binaries at `usr/libexec/docker/cli-plugins/docker-compose` and `usr/libexec/docker/cli-plugins/docker-buildx` that trigger trivy at the moment. I don't think we need them, how about we just remove those from the image to silence the alerts?

**Comment by willdollman** [comment]
Is the image just going to be used by this one customer, or all customers with k8s + executors?

If the image hosted by us is just going to be used by a single customer, it's less of an issue about vulns as we can set the customer's expectation that we're just rehosting the image.
But if the deployment config will be updated so that all customers use the sourcegraph-hosted image, we take on (in the customer's eye) more responsibility for keeping the image updated and patched. We'd likely receive feedback from customers like @alexfogg mentioned requesting us to remove vulns, and as we don't have much control over the image we wouldn't be able to do much about it.

@eseliger's idea of removing unneeded vulnerable binaries is a good one, and we should do that regardless. Though in the future we still might see vulns identified in binaries that we can't remove.

**Comment by Piszmog** [comment]
@davejrt Thank you for pushing those changes to remove the unneeded dependencies!

File: docker-images/dind/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM docker:23.0.1-dind@sha256:ed6220b0de0f309f0844cf8cf1a6b861e981fb7f5c28bec6acc97abc910bd0a8
+
+RUN rm -f /usr/libexec/docker/cli-plugins/docker-compose && \
+    rm -f /usr/libexec/docker/cli-plugins/docker-buildx
+
+ARG COMMIT_SHA="unknown"
+ARG DATE="unknown"
+ARG VERSION="unknown"
```
Comment (eseliger) [inline docker-images/dind/Dockerfile:8]:
```suggestion
ARG COMMIT_SHA="unknown"
ARG DATE="unknown"
ARG VERSION="unknown"

LABEL org.opencontainers.image.revision=${COMMIT_SHA}
LABEL org.opencontainers.image.created=${DATE}
LABEL org.opencontainers.image.version=${VERSION}
LABEL com.sourcegraph.github.url=https://github.com/sourcegraph/sourcegraph/commit/${COMMIT_SHA}
```

Comment (Piszmog) [reply docker-images/dind/Dockerfile:8]: Should we also include `org.opencontainers.image.url`, `org.opencontainers.image.source`, `org.opencontainers.image.documentation`?

Comment (eseliger) [reply docker-images/dind/Dockerfile:8]: Maybe, those ones I suggested are the ones we set for our other images, I think our renovate tooling makes use of them

Comment (Piszmog) [reply docker-images/dind/Dockerfile:8]: I'll add them just in case

**Comment by Piszmog** [comment]
If I can get one more sanity check, I'll merge this. Once merged I'll update the trivy image list

**Comment by eseliger** [comment]
once merged, let's also update the helm chart etc.

**Review by sanderginn — APPROVED** [review]

**Comment by malomarrec** [comment]
We're just one security review away of making a user's life much better!

**Comment by evict** [comment]
There were some more CVEs present pushed a commit that resolves that:

```
$ grype sha256:62a8c5795f6c5ca02f857a290c8174af95eee1f719e90f1ab5938137a443325a 
 ✔ Vulnerability DB        [no update available]
 ✔ Loaded image            
 ✔ Parsed image            
 ✔ Cataloged packages      [64 packages]
 ✔ Scanning image...       [0 vulnerabilities]
   ├── 0 critical, 0 high, 0 medium, 0 low, 0 negligible
   └── 0 fixed
No vulnerabilities found
```

**Review by evict — APPROVED** [review]

**Comment by Piszmog** [comment]
Thank you all for helping to get this merged!

Additional PRs will come for trivy and helm

**Comment by malomarrec** [comment]
@Piszmog @eseliger can I go ahead and communicate to the customer that the image is available?

**Comment by Piszmog** [comment]
@malomarrec yep!

**Comment by eseliger** [comment]
https://github.com/sourcegraph/sourcegraph/pull/49062 Was just merged so should be good

**Pull request merged**